crates/interpreter/src/exec.rs

```diff
@@ -1036,24 +1044,47 @@ impl<'m> Thread<'m> {
         self.values().split_off(len)
     }
 
-    fn push_label(&mut self, type_: FuncType<'m>, kind: LabelKind) {
+    fn push_label(&mut self, type_: FuncType<'m>, kind: LabelKind, val_cnt: Option<u32>) {
         let arity = match kind {
             LabelKind::Block | LabelKind::If => type_.results.len(),
             LabelKind::Loop => type_.params.len(),
         };
-        let label = Label { arity, values_cnt: type_.params.len() };
+        match kind {
+            LabelKind::If => {
+                if val_cnt.is_some() {
+                    debug_assert_eq!(arity as u32, val_cnt.unwrap())
+                }
+            }
+            LabelKind::Loop => {
+                // There is no side table entry for a loop that is only executed once.
+                if val_cnt.is_some() && !type_.params.is_empty() {
+                    debug_assert_eq!(arity as u32, val_cnt.unwrap())
+                }
+            }
+            _ => {}
+        }
+        let label = Label { arity, values_cnt: type_.params.len(), kind };
         self.label().values_cnt -= label.values_cnt;
         self.labels().push(label);
     }
 
-    fn pop_label(mut self, l: LabelIdx, offset: usize) -> ThreadResult<'m> {
+    fn pop_label(mut self, l: LabelIdx, offset: usize, pop_cnt: Option<u32>) -> ThreadResult<'m> {
         let i = self.labels().len() - l as usize - 1;
         if i == 0 {
             return self.exit_frame();
         }
         let frame = self.frame();
         let values_cnt: usize = frame.labels[i ..].iter().map(|label| label.values_cnt).sum();
-        let Label { arity, .. } = frame.labels.drain(i ..).next().unwrap();
+
+        let Label { arity, kind, .. } = frame.labels.drain(i ..).next().unwrap();
+        match kind {
+            LabelKind::Block => {}
```

Comment:
There is no side table entry for `Block`, but `push_label()` and `pop_label()` in `exec.rs` process `Block`. Do we have to keep `Label` just for the sake of `Block` (we could remove `Label` if `Block` is no a `Label` in `exec.rs`)?

Reply:
Actually even before trying to understand this question, I'm confused about the code. `val_cnt` and `pop_cnt` should only matter for jump instructions (`Br`, `BrIf`, `BrTable`, `Else`).

Reply:
Made the change to check `val_cnt` in `pop_label` instead.

Reply:
A little surprisingly, the `val_cnt` and `pop_cnt` computation in `valid.rs` seems perfect.

Reply:
The debug assert should hold for all kinds of labels, since we can pop out of any of them. `Block` should not be special. Does the debug assert fail in that case? If yes this probably indicates a bug somewhere.

Reply:
The side table index should be `offset` instead 0. Will post a PR in the Google directory.